draft-ietf-quic-transport.md

```diff
@@ -2675,201 +2676,289 @@ created, it MUST close the connection with error code STREAM_STATE_ERROR.
 
+~~~
+       o
+       | Application Open
+       | Open Paired Stream (bidirectional)
+       v
+   +-------+
+   | Open  | Send RST_STREAM
+   |       |-----------------------.
+   +-------+                       |
+       |                           |
+       | Send STREAM /             |
+       |      STREAM_BLOCKED       |
+       v                           |
+   +-------+                       |
+   | Send  | Send RST_STREAM       |
+   |       |---------------------->|
+   +-------+                       |
+       |                           |
+       | Send STREAM + FIN         |
+       v                           v
+   +-------+                   +-------+
+   | Data  | Send RST_STREAM   | Reset |
+   | Sent  +------------------>| Sent  |
+   +-------+                   +-------+
+       |                           |
+       | Recv All ACKs             | Recv ACK
+       v                           v
+   +-------+                   +-------+
+   | Data  |                   | Reset |
+   | Recvd |                   | Recvd |
+   +-------+                   +-------+
+~~~
+{: #fig-stream-send-states title="States for Send Streams"}
 
-### half-closed (remote) {#state-hc-remote}
+The sending part of stream that the endpoint initiates (types 0 and 2 for
+clients, 1 and 3 for servers) is opened by the application or application
+protocol.  The "Open" state represents a newly created stream that is able to
+accept data from the application.  Stream data might be buffered in this state
+in preparation for sending.
 
-A stream is "half-closed (remote)" when the stream is no longer being used by
-the peer to send any data.  An endpoint will have either received all data that
-a peer has sent or will have received a RST_STREAM frame and discarded any
-received data.
+The sending part of a bidirectional stream initiated by a peer (type 0 for a
+server, type 1 for a client) enters the "Open" state if the receiving part
+enters the "Recv" state.
 
-Once all data has been either received or discarded, a sender is no longer
-obligated to update the maximum received data for the connection.
+Sending the first STREAM or STREAM_BLOCKED frame to be sent causes a send stream
+to enter the "Send" state.  An implementation might choose to defer allocating a
+Stream ID to a send stream until it sends the first frame and enters this state,
+which can allow for better stream prioritization.
 
-Due to reordering, an endpoint could continue receiving frames for the stream
-even after the stream is closed for sending.  Frames received after a peer
-closes a stream SHOULD be discarded.  An endpoint MAY choose to limit the period
-over which it ignores frames and treat frames that arrive after this time as
-being in error.
+In the "Send" state, an endpoint transmits - and retransmits as necessary - data
+in STREAM frames.  The endpoint respects the flow control limits of its peer,
+accepting MAX_STREAM_DATA frames.  It generates STREAM_BLOCKED frames if it
```

Comment:
You talk about flow control a fair amount here.  Is it worth adding a 'blocked' state to the state machine?  I'm also happy with this as is and adding it later if we think it's useful.  Just a thought.

Reply:
You'd have to have it sit to the side of Send, though, with the ability to transition from Open to Blocked (no credit to start with) and the ability to oscillate between Blocked and Send, *and* the ability to close or reset from there as well.  It feels like it would complicate the concept of the state machine much more than it would clarify.

Just say that, while in the Send state, you MUST obey flow control, sending STREAM_BLOCKED if you're stuck.

Reply:
Yeah, blocked is a sub-state that doesn't really add much at this level of the description.